bokehjs/src/lib/embed/standalone.ts

```diff
@@ -11,15 +11,17 @@ import {BOKEH_ROOT} from "./dom"
 export const index: {[key: string]: View} = {}
 
 export async function add_document_standalone(document: Document, element: HTMLElement,
-    roots: {[key: string]: HTMLElement} = {}, use_for_title: boolean = false): Promise<View[]> {
+    roots: HTMLElement[] = [], use_for_title: boolean = false): Promise<View[]> {
   // this is a LOCAL index of views used only by this particular rendering
   // call, so we can remove the views we create.
   const views: Map<HasProps, View> = new Map()
 
   async function render_model(model: HasProps): Promise<View> {
     let root_el: HTMLElement
-    if (model.id in roots)
-      root_el = roots[model.id]
+    const root_models = document.roots()
+    const idx = root_models.indexOf((model as any))
+    if ((idx < roots.length) && (roots[idx] !== null))
```

Comment:
This is an unfortunate part of the TypeScript's type system. One can perform `== null` (or `!= null`) even if the type doesn't mention `null` (under strict null checks). So, you code is correct from run time perspective, but you won't be able to pass a `null` from type system's perspective. You will have to change the type to `(HTMLElement | null)[]`. I will implement a lint rule to disallow code like this at some point.

Reply:
It's also my preference to use `!=` instead of `!==`, unless it's your intent to distinguish between `null` and `undefined`. Though then you would indicate this in the type system by using ` | null | undefined` (or some other variant).

Reply:
I will look at this overall change once more today, because I think this could be simplified further.

Reply:
That would be appreciated. It does seem like something like this approach will work though.

Reply:
If I change the typing to `(HTMLElement | null)[]` I get this error despite the `roots[idx] != null` check:

```
    src/lib/embed/standalone.ts:24:7 - error TS2322: Type 'HTMLElement | null' is not assignable to type 'HTMLElement'.
      Type 'null' is not assignable to type 'HTMLElement'.

    24       root_el = roots[idx]
```

Reply:
Seems like it can't handle the indexing part.